Getting started with spring boot (#450)
reflectoring/reflectoring.github.io: content/blog/2022/2022-01-18-getting-started-with-spring-boot.md

The new approach helps us to start the development process much faster.

It is important to note that the configurations that come with the `starter` dependencies are not invasive. We can create custom configurations only where we need them.
## Building The Database Entities
{{% image alt="Spring Boot High-level arch " src="images/posts/spring-boot-begginer-guide/SpringApplication-entity.png" %}}

The database entity represents the direct link with the database tables. Entity classes represent columns, relationships between different tables, and constraints.

While creating database entities, we have to think about requirements from the beginning of the article:
- The user has to provide their name, last name, email, and password.
- Each book has information about the author, date of publication, number of instances, and users that currently own the book.
- The user can see all books and borrow them.

After sketching which data tables need to contain, we will create those objects in the Java code.

### Sketching the Database Entities

> We will build the database diagram for our application:

Reviewer: We're not building a diagram, we're building software! :)

### Defining a Primary Key
```java
@Entity(name = "book")
public class Book {

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private long id;
    //The rest of the class omitted

}
```
When defining a class as an entity, we need to provide the id column. The id column will be the primary key of that table.

We need to decide how will the id column be generated. Using the `@GeneratedValue` annotation we can define several different strategies:
- [IDENTITY](#identity-generation-strategy)
- [SEQUENCE](#sequence-generation-strategy)
- [TABLE](#table-generation-strategy)
- [AUTO](#defining-the-primary-keys)

We used the `AUTO` strategy, which leverages whatever the database prefers. Most databases prefer to use the `SEQUENCE` strategy for their primary key definition.

#### Identity Generation Strategy
The `GenerationType.IDENTITY` strategy allows the database to autoincrement the id value when we insert a new row.
Let us look into the example of using the identity generation strategy:
```java
@Entity(name = "book")
public class Book {

    @Id
    @GeneratedValue( strategy = GenerationType.IDENTITY)
    private long id;
    
    // Rest of the code omitted
}
```

> We define the identity generation type in the strategy attribute of the `@GeneratedValude`. While the identity strategy is highly efficient for the database, it doesn't perform well with the Hibernate ORM.

Reviewer: ```suggestion
We define the identity generation type in the strategy attribute of the `@GeneratedValue`. While the identity strategy is highly efficient for the database, it doesn't perform well with the Hibernate ORM.
```

> The Hibernate expects that every managed entity has its id set, so it needs to go and call the database to insert the id.

Reviewer: ```suggestion
Hibernate expects that every managed entity has its id set, so it needs to go and call the database to insert the id.
```

> The Hibernate expects that every managed entity has its id set, so it needs to go and call the database to insert the id.

Reviewer: What is Hibernate? This is the first time it's mentioned and as a newbie I'm confused :). Maybe mention it in the section about the JPA dependency above. Also explain what JPA stands for, while you're at it.

The identity strategy is excellent for fast iteration and the early stages. When we move to the development environment, we should move to something more stable and with better performance.

#### Sequence Generation Strategy

> The `GenerationType.SEQUENCE` strategy benefits from using the sequences inside the databases:

Reviewer: The strategy doesn't benefit from anything. This sentence makes no sense :). Please rephrase.

```java
@Entity(name = "book")
public class Book {

    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE, 
            generator = "book_generator")
    @SequenceGenerator(name = "book_generator",
            sequenceName = "book_seq",
            initialValue = 10)
    private long id;
}
```

> To define the sequence generator, we annotate the field with the `@SequenceGenerator`. We declare the name for the Hibernate, the sequence name, and the initial value.

Reviewer: ```suggestion
To define the sequence generator, we annotate the field with the `@SequenceGenerator` annotation. We declare the name for Hibernate, the sequence name, and the initial value.
```

After defining the generator, we need to connect it to the `@GeneratedValue` annotation by setting its name in the `generator` attribute. 

The sequence strategy uses the sequences from the databases to determine which primary key to chose next.

#### Table Generation Strategy

> Similar to the sequence generator, the table strategy uses the table to keep track of which primary key can be next:

Reviewer: This sentence doesn't make sense.

```java
@Entity(name = "book")
public class Book {

    @Id
    @GeneratedValue(strategy = GenerationType.TABLE,
            generator = "book_generator")
    @TableGenerator(name = "book_generator", table = "book_id_table")
    private long id;
}
```
We need to define the `@TableGenerator` annotation with the name and table attributes.
We provide the generator name to the `generator` attribute inside the `@GeneratedValue` annotation.

> The table generator uses

Reviewer: Fell asleep while writing? :)

### Defining the Column

> We define the table column with the `@Column` annotation and the name of the column inside the `name` attribute:

Reviewer: ```suggestion
We define each table column with the `@Column` annotation and the name of the column inside the `name` attribute:
```

```java
@Entity(name = "book")
public class Book {

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private long id;

    @Column(name = "title")
    private String title;

    @Column(name = "author")
    private String author;

    @Column(name = "publication")
    private Date publication;

    @Column(name = "numberOfInstances")
    private int numberOfInstances;

    //The rest of the class omitted

}
```

> If we do not provide the `name` attribute, the Hibernate will assume that the column name is the same as the variable name inside the Java class.

Reviewer: ```suggestion
If we do not provide the `name` attribute, Hibernate will assume that the column name is the same as the field name inside the Java class.
```

It is always better to be safe and set the `name` attribute so that things do not start crashing when someone accidentally changes the variable name.

### Defining the ManyToMany Relationship

> We define the many-to-many relationship with the `@ManyToMany` annotation

Reviewer: ```suggestion
We define the many-to-many relationship between `Book` and `User` with the `@ManyToMany` annotation
```
It's a good idea to always repeat some context, especially if there were a few sections between introducing a concept and now.

Each relationship has two sides:
- the owner - the table that has the foreign key
- the target - the table to which one the owner is refering to with foreign key

#### The Owner of the Relationship
One side of that relationship needs to be the owning side and define the join table.

> We decided that it will be the User side:

Reviewer: ```suggestion
We decided that it will be the `User` side:
```

    @ManyToMany(mappedBy = "borrowedBooks")
    private List<User> users;

    // Rest of the code omitted
}
```
We defined the target side with the `@ManyToMany` annotation `mappedBy` attribute. We set the `mappedBy` attribute to the name on the owning side.

### Configuring the Database
We said that the entity class represents the direct link to the table in the database.
We set the configuration in Spring Boot through the `application.properties` file.

Even though the Spring Boot framework comes with the configurations for most things, we need to tap in and change them.
The excellent thing about provided configurations is that they are non-invasive, and we can change only those things that we need.

In this example, we will show how to configure the database. The H2 database can be an in-memory or persistent file-based database, and we will show how to set them up.

#### Defining the Database URL
Let us look how we define the URL for the in-memory database.
```text
spring.datasource.url=jdbc:h2:mem:localdb
spring.jpa.database-platform=org.hibernate.dialect.H2Dialect
spring.datasource.driver-class-name=org.h2.Driver
# Rest of the configuration is omitted
```

> After defining the keywords `jdbc` and the `h2`to note that we are using the H2 database, we are defining that we are using the in-memory database with the `mem` keyword.

Reviewer: ```suggestion
After defining the keywords `jdbc` and `h2` to note that we are using the H2 database, we are defining that we are using the in-memory database with the `mem` keyword.
```

The `localdb` is the database name and can be whatever we want.

The in-memory database is good for fast iterations and prototyping, but we need something more persistent when we go into full development.
Let us see how to define the file-based H2 database:

```text
spring.datasource.url=jdbc:h2:file:/Users/mateostjepanovic/Documents/git/code-examples/spring-boot/begginers-guide/src/main/resources/data/demo;AUTO_SERVER=true
spring.jpa.database-platform=org.hibernate.dialect.H2Dialect
spring.datasource.driver-class-name=org.h2.Driver
# Rest of the configuration is omitted
```
After defining the `jdbc` and the `h2` keywords, we note that we want to use a file-based H2 database with the `file` keyword. The last part of the URL is the absolute path to the folder where we want to save our database.

{{% info title="The H2 database" %}}

> Please note that we should use the H2 only for the development phase. When going to the production environment, move to something more production-ready ( Oracle, PostgreSQL, etc.).

Reviewer: ```suggestion
Please note that we should use the H2 database only for the development phase. When going to the production environment, move to something more persistent and production-ready like Oracle, PostgreSQL, etc.
```

{{% /info %}}

#### Defining the Database Login Information
After we defined the URL for the database we need the login informations for that database:

```text
spring.datasource.username=username
spring.datasource.password=password
spring.h2.console.enabled=true

# Rest of the configuration is omitted
```
We define the username and password for the database.

Please define that the password and username should be externalized and encrypted in the production environment.

More about externalizing configurations can be found in the [Build Once, Run Anywhere: Externalize Your Configuration](https://reflectoring.io/externalize-configuration/) article.

#### Defining Schema Creation
We can set the configuration to autocreate the schema for us:
```text
spring.jpa.hibernate.ddl-auto=create
spring.jpa.generate-ddl=true

# Rest of the configuration is omitted
```

> The Hibernate DDL is used to generate the schema using the entity definitions. With the `ddl-auto`, we set that we want to destroy and recreate the schema on each run.

Reviewer: ```suggestion
Setting `hibernate.ddl-auto` to `create`, we tell Hibernate that we want to destroy and recreate the schema on each run. Only use this setting for testing!
```

## Building the Data Access Layer

{{% image alt="Spring Boot High-level arch " src="images/posts/spring-boot-begginer-guide/SpringApplication-dataAccessLayer.png" %}}

After creating entities, we can develop the data access layer. The data access layer allows us to use methods to manipulate the data in the database. We will build the data access layer using the repository pattern.

The repository pattern is the design pattern that leverages the usage of interfaces for connection to the database. The repository interface hides all implementation details of connecting to the database, maintaining the connection, transactions, etc.

We will show how to create the repository for the entity and explain how we use it.

### Creating the Book Repository
```java
public interface BookRepository extends JpaRepository<Book, Long> {
   // Rest of the code is omitted 
}
```

> Extending `JpaRepository` turns our interface into the repository bean for the application context to control. 

Reviewer: What is the application context? This is the first time this word is used. As a newbie, I'm confused. Maybe add a short section about the Spring programming model after the "Setting up the project" section? That section should explain the application context, what a bean is, what autowiring is and how it works. Maybe  link to https://reflectoring.io/constructor-injection/

Just realized that you have a section like this in "Injecting the Repository Class". Good! Mention that the application context is explained in that section, then!

The primary job of the application context is to manage beans lifecycle.

> When a class is a Spring Bean we are sure that we will get the instance the class where ever it is asked using the `@Autowired` annotation.

Reviewer: ... or one of the other ways of autowiring / injection (mention the upcoming section about  injecting the Repository Class).

#### Field-based Injection
Let us look at how to use the field-based injection:
```java
@Service
public class GetBookService {
    @Autowired
    private final BookRepository bookRepository;

    // Rest of the code is omitted
}
```

Spring recognizes the `@Autowired` annotation and makes sure that the `BookRepository` is provided.
#### Setter-based Injection
```java
@Service
public class GetBookService {

    private BookRepository bookRepository;

    @Autowired
    private void setBookRepository(BookRepository bookRepository){
        this.bookRepository = bookRepository;
    }

    // Rest of the code is omitted
}
```

We can put the `@Autowired` annotation on the setter method. The setter-based injection does not allow us to mark the variable as final.
#### Constructor-based Injection
```java
@Service
public class GetBookService {

    private final BookRepository bookRepository;

    @Autowired
    public GetBookService(BookRepository bookRepository) {
        this.bookRepository = bookRepository;
    }

    // Rest of the code is omitted
}
```
For the constructor-based annotation, we should set the `@Autowired` annotation on the constructor. 

> We are sure that the IoC container will provide all constructor arguments.

Reviewer: What does IoC mean? Please explain in the parent section.

More about dependency injection and why to use the constructor-based injection we can find in the [article.](https://reflectoring.io/constructor-injection/)

## Building Controllers
After creating entities, repositories, and services let us create our first controller that is solving the requirement that the application should be accessable through the browser.

{{% image alt="Spring Boot High-level arch " src="images/posts/spring-boot-begginer-guide/SpringApplication-webLayer.png" %}}

The imported `spring-boot-starter-web` contains everything that we need for the controller. We have Spring MVC autoconfigured and the local server ready to use.

With Spring MVC, we can define a controller with the `@Controller` or `@RestController` annotation so it can handle incoming requests.

> ### Creating an Endpoint With the @RestController

Reviewer: ```suggestion
### Creating an Endpoint With @RestController
```

We are going to create our first endpoint. We want to fetch all books that the bookstore owns. This is the endpoint that will be called from the homepage.

Let us look into the codebase:
```java
@RestController
@RequestMapping("/books")
public class BooksRestController {

    private final GetBookService getBookService;

    @Autowired
    public BooksRestController(GetBookService getBookService) {
        this.getBookService = getBookService;
    }

    @GetMapping
    List<BookResponse> fetchAllBooks(){
        return getBookService.getAllBooks();
    }

    // Rest of the code 
}
```
To make our class the controller bean we need to annotate it with `@RestController` or with `@Controller`.

> The difference between these two is that `@RestController` automatically wraps the object into `ResponseEntity<>`. 

Reviewer: Which object? The return value of all methods that answer to a request, i.e. all methods that are annotated with `@GetMapping` etc.
